query/src/main/java/org/infinispan/query/concurrent/InfinispanIndexingExecutorProvider.java

```diff
@@ -0,0 +1,64 @@
+package org.infinispan.query.concurrent;
+
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.hibernate.search.backend.lucene.work.spi.LuceneWorkExecutorProvider;
+import org.hibernate.search.engine.common.execution.spi.SimpleScheduledExecutor;
+import org.infinispan.factories.AbstractComponentFactory;
+import org.infinispan.factories.AutoInstantiableFactory;
+import org.infinispan.factories.annotations.DefaultFactoryFor;
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.factories.scopes.Scope;
+import org.infinispan.factories.scopes.Scopes;
+import org.infinispan.util.concurrent.BlockingManager;
+
+@Scope(Scopes.GLOBAL)
+@DefaultFactoryFor(classes = LuceneWorkExecutorProvider.class)
+public class InfinispanIndexingExecutorProvider extends AbstractComponentFactory implements AutoInstantiableFactory, LuceneWorkExecutorProvider {
+
+   @Inject
+   BlockingManager blockingManager;
+
+   @Override
+   public Object construct(String name) {
+      return new InfinispanIndexingExecutorProvider();
+   }
+
+   @Override
+   public SimpleScheduledExecutor writeExecutor(Context context) {
+      return new InfinispanScheduledExecutor(blockingManager);
+   }
+
+   static final class InfinispanScheduledExecutor implements SimpleScheduledExecutor {
+
+      private final BlockingManager blockingManager;
+
+      public InfinispanScheduledExecutor(BlockingManager blockingManager) {
+         this.blockingManager = blockingManager;
+      }
+
+      @Override
+      public Future<?> submit(Runnable task) {
+         return blockingManager.runBlocking(task, this).toCompletableFuture();
+      }
+
+      @Override
+      public ScheduledFuture<?> schedule(Runnable command, long delay, TimeUnit unit) {
+         return blockingManager.scheduleRunBlocking(command, delay, unit, this);
+      }
+
+      @Override
+      public void shutdownNow() {
+         // The executors lifecycle is handled at global Infinispan level
+      }
+
+      @Override
+      public boolean isBlocking() {
+         // the task (Runnable) can be blocking but the blockingManager will accept all the tasks
+         // and will sooner produce OOM rather than block
```

Comment:
It doesn't produce OOM or accept all tasks.

1. If the invoking thread is a non blocking thread and the blocking pool and its queue are full then it will throw an exception
2. If the invoking thread is not a non blocking thread and the blocking pool and its queue are full then it will run the task immediately in the invoking thread.

In either case the invocation is not "blocking" but it may not return immediately as it is performing the task. I am not 100% sure if the 2 above is okay or not though.

Reply:
If it is, I would just reword the comments here

Reply:
I think it is OK. In any case we're probably going to use non blocking threads soon when we invoke it as soon as we address [the ISPN-14483](https://issues.redhat.com/browse/ISPN-14483). I fixed the comment here. Thank you